### DISCUSSION TITLE
Helper Help

### DISCUSSION THREAD

**Discussion by aljawaid** [aljawaid/ContentCleaner#3] [Q&A] [answered]
**Helper:**
```php
<?php

namespace Kanboard\Plugin\ContentCleaner\Helper;

use Kanboard\Core\Base;

class CleaningHelper extends Base
{
    public function countTablesDB()
    {
        $db_size = $this->applicationCleaningModel->countTables();

        return $db_size;
    }
}
```

**Model:**
```php
class ApplicationCleaningModel extends Base
{
    public function countTables()
    {
        //return $this->db->execute('SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = `myppworkspace`');

        $table_schema = $this->db->table(self::TABLE);
        return $this->db->execute('SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = `'. $table_schema .'`');
    }
}
```

**Calling:**
`<?= $this->helper->cleaningHelper->countTablesDB() ?>`

I am getting error:
`Uncaught Error: Undefined class constant 'TABLE'`

**Comment by aljawaid** [comment, marked as answer]
@creecros could you please have a look?

**Reply by aljawaid** [reply]
I am guessing this is incorrect: `$table_schema = $this->db->table(self::TABLE);`

**Reply by aljawaid** [reply]
so in human terms it should be self::database... let me search the kanboard code for that

**Reply by aljawaid** [reply]
```php
$table_schema = DB_NAME;
        return $this->db->execute('SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = `'. $table_schema .'`');
```

is giving me `Internal Error: SQL Error: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'myppworkspace' in 'where clause'`

I think I am on the right path

**Reply by creecros** [reply]
`TABLE` in that context is a constant. did you define that constant?

**Reply by aljawaid** [reply]
This is the SQL query which retrns the number of tables.... providing my helper/model is correct (please check above)
`SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = 'myppworkspace' AND TABLE_TYPE = 'BASE TABLE`

In the model it is like this:
```
return $this->db->execute('SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = '. DB_NAME .' AND TABLE_TYPE = "BASE TABLE";');
```

In phpmyadmin, I get 59 tables
In kanboard, I get `Internal Error: SQL Error: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'myppworkspace' in 'where clause'`

**Reply by aljawaid** [reply]
Ultimately, I am trying to display the database size and tables count.... this one is for the tables

**Reply by aljawaid** [reply]
This seems wrong:
```
return $this->db->execute('SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = '. DB_NAME .' AND TABLE_TYPE = "BASE TABLE";');
```
even if I remove the semicolon... I think the quotes are wrong but I'm not sure how to combine them

**Reply by creecros** [reply]
i suck at mysql statements, so i truly do not know for sure. I would say, a couple of things though, to keep in mind. taking this approach, your statment might only work for mysql, and maybe not for postgres, sqlite, etc. To avoid, i would not perform sql statements. I would use picodb without specific sql statements. that way you dont have to worry about those issues.

as for the statment you have, im clueless. i feel like you are missing quotes though, maybe i am wrong:
```
return $this->db->execute('SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = "'. DB_NAME .'" AND TABLE_TYPE = "BASE TABLE";');
```

again, i would avoid this anyway.

**Reply by creecros** [reply]
Example:

```
$count = $db->table('information_schema.tables')
            ->where('table_schema', '=', 'mydatabase')
            ->where('TABLE_TYPE', '=', 'BASE TABLE')
            ->count();
```

**Reply by aljawaid** [reply]
Cool thanks for that example, I will try it shortly. I agree with you and thought about picodb but I couldn't find a part for the full dB. Your example looks like a good starting point so I will check that out.

Would the result be `return $count`?

**Reply by creecros** [reply]
theoretically, it should be the 59, right?
i could be wrong on the where() method. worth a try though. if not, i know ->eq() works

dont hold me to this, you may have been on the correct path, just explore it more otherwise you will have to account for all db types. im litteraly going off the top of my head on a crappy memory at the moment

**Reply by aljawaid** [reply]
yes, my result should be 59... I will try it out using `return $count`

**Reply by aljawaid** [reply]
This seems to give me 59 without any errors... I think we have got somewhere on the table count

```php
public function countTables()
    {
        //return $this->db->execute('SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = `myppworkspace`');
        //SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = 'myppworkspace' AND TABLE_TYPE = 'BASE TABLE';

        //$table_schema = DB_NAME;
        //return $this->db->execute('SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = `'. DB_NAME .'` AND TABLE_TYPE = `BASE TABLE`');

        return $this->db->table('information_schema.tables')
            ->eq('table_schema', 'myppworkspace')
            ->eq('TABLE_TYPE', 'BASE TABLE')
            ->count();
    }
```

**Comment by creecros** [comment]
So, this is promising. 

```
<?php

$file = __DIR__.'/../../../../app/Schema/Mysql.php'; // replace with your file path

$tables = array();
$columns = array();

$sql = file_get_contents($file);

// extract table names
preg_match_all("/CREATE\s+TABLE\s+`?(\w+)`?/i", $sql, $matches);
$tables = $matches[1];

// extract column names
preg_match_all("/ALTER\s+TABLE\s+`?(\w+)`?.*ADD\s+`?(\w+)`?/i", $sql, $matches);
foreach ($matches[1] as $i => $table) {
    $column = $matches[2][$i];
    $columns[$table][] = $column;
}

// extract renamed table names
preg_match_all("/ALTER\s+TABLE\s+`?(\w+)`?\s+RENAME\s+TO\s+`?(\w+)`?/i", $sql, $matches);
foreach ($matches[1] as $i => $old_table) {
    $new_table = $matches[2][$i];
    // update table name in $tables array
    $key = array_search($old_table, $tables);
    if ($key !== false) {
        $tables[$key] = $new_table;
    }
    // update table name in $columns array
    if (isset($columns[$old_table])) {
        $columns[$new_table] = $columns[$old_table];
        unset($columns[$old_table]);
    }
}

// extract dropped table names
preg_match_all("/DROP\s+TABLE\s+`?(\w+)`?/i", $sql, $matches);
foreach ($matches[1] as $table) {
    // remove table name from $tables array
    $key = array_search($table, $tables);
    if ($key !== false) {
        unset($tables[$key]);
    }
    // remove table name from $columns array
    if (isset($columns[$table])) {
        unset($columns[$table]);
    }
}

// print the results
print_r(count($tables).'<br>');
foreach ($tables as $table) { print_r($table.'<br>'); }
//print_r(count($columns));

?>
```

It comes up with 47 Tables:
```
47
sessions
predefined_task_descriptions
invites
column_has_restrictions
project_role_has_restrictions
project_has_roles
column_has_move_restrictions
tags
task_has_tags
project_has_files
task_has_external_links
password_reset
project_has_groups
groups
group_has_users
user_has_metadata
project_has_metadata
task_has_metadata
project_has_notification_types
user_has_unread_notifications
user_has_notification_types
custom_filters
plugin_schema_versions
project_daily_stats
currencies
transitions
links
task_has_links
subtask_time_tracking
swimlanes
project_daily_summaries
project_activities
settings
user_has_notifications
task_has_subtasks
task_has_files
project_has_categories
remember_me
last_logins
users
projects
project_has_users
columns
tasks
comments
actions
action_has_params
```

**Reply by aljawaid** [reply]
This might be useful but I think its not PicoDB?

**Reply by creecros** [reply]
nah, that's parsing the schema to get the tables that KB should have. so, if you have 59, and kanbaord should have 47, now you just compare the two, and you can find out which tables are not standard tables. then you can say, these 12 tables have been added by plugins...59 - 47 = 12.

then you take it further, run that code on schemas for the installed plugins, add those up, and filter put what's left, and you know exactly which tables you don't need.

**Reply by creecros** [reply]
I'm still working on breaking each table down with respective columns, that's a bit tougher, haven't got there yet.

**Reply by aljawaid** [reply]
before I get started on this bit... this code above, is it for a standalone php page or can I just copy the whole lot into a model and create a helper?

I think, for this plugin, it will be good to count the core tables (as your code above) and display them discretly too, so your above efforts wont be wasted as users of this plugin would deffo find it useful.

i could just do an `if` to target the relevant schema file

**Reply by creecros** [reply]
should work fine in a helper. that said, it worked for me if I used the postgres and mysql, only. but it shouldn't matter if you check the mysql schema, and you are using sqlite, because no KB release has different tables for different dbs. it really comes down to the version of KB when you would see different tables. I would say, it would be wise to confirm it, and I haven't done that, as of yet.

**Reply by aljawaid** [reply]
you mean just use it in a helper..... without a model?

**Reply by creecros** [reply]
don't see why you would need a model

**Reply by aljawaid** [reply]
> don't see why you would need a model
cool, I will copy it as it is above into a helper.  I asked because I remember you once said to access the database you must go through the model.

**Reply by creecros** [reply]
you aren't accessing the db.

**Reply by aljawaid** [reply]
ok, interesting to know... although that confuses me abit.

**Reply by creecros** [reply]
how so? ` ..\Schema\Mysql.php` is not the db, it is the file used to build the db.

**Reply by creecros** [reply]
a better explanation would be, its no different than you opening that file yourself, reading it with your own eyes, and writing down the name of every table you see.

**Reply by aljawaid** [reply]
That makes a lot more sense now.... we are just reading the file containing database setup code.... we are not actually using the database.  I understand now.  Thanks

**Reply by aljawaid** [reply]
https://github.com/aljawaid/ContentCleaner/blob/0d9e8afc1696de0217626042e3adf5208cc5fa36/Helper/DefaultTableHelper.php#L24

This works for me, but is it correct?  If I use your original suggestion of `__DIR__` then the path includes the `plugins` folder and plugin folder

**Reply by creecros** [reply]
__DIR__ is the base path of where the code is, so when I was testing it, it was valid for where it was at the time. it will be different if it is in a helper now.
